### PR TITLE
Problem: When packages reside in un-indexed channels, the test command fails

### LINKED ISSUE #2301: When testing a package, the folder containing the package needs be an indexed channel
When testing packages using: `conda-build /data/builds/linux-64/eigen3-3.2.9-he32ef9e_5.tar.bz2 --test` and the artifact is just in a folder without repodata.json, the above command would produce: 

```
conda.exceptions.ResolvePackageNotFound: 
  - eigen3 3.2.9 he32ef9e_5
```

after running `conda index` on that folder the above test is run successfully...

Going back to our discussion of #2260, where we concluded that testing recipes is unintuitive: I'd say that its also unintuitive to have to put a package-under-test (that we even specify by full path) into a valid conda channel... 

It should be: *CB, please test that package there!*, not *CB, please test that package in that channel there*.

### PR DESCRIPTION
**Pull request by gabm**
This fixes #2301 

@msarahan Please review the fix carefully. There is more logic involved to index the right channels.. Maybe at some point we could teach update_index to do globbing...

There is one point where I'm not sure, and its where I throw the error... I'll mark it in the code

### CODE REVIEW

File: conda_build/build.py [hunk 1]
```diff
@@ -1347,24 +1347,39 @@ def _construct_metadata_for_test_from_package(package, config):
...
+    # get last part of the path
+    last_element = os.path.basename(local_pkg_location)
+    is_channel = False
+    for platform in ('win-', 'linux-', 'osx-', 'noarch'):
+        if last_element.startswith(platform):
+            is_channel = True
+
+    if not is_channel:
+        log.error("Your package must reside in a channel structure with platform-subfolders!")
+        sys.exit(1)
```
Comment (gabm) [inline conda_build/build.py]: here I'm not sure how to properly throw an error?!

Comment (msarahan) [reply conda_build/build.py]: ```raise ValueError("Your package must reside in a channel structure with platform-subfolders!")```

A more specific exception type might be nice, but ValueError is probably fine to start with.

Comment (gabm) [reply conda_build/build.py]: done...

**Comment by msarahan** [comment]
Thanks!  I'll wait on CI and merge if they come out OK.  Looks fine to me.

**Comment by msarahan** [comment]
Looks great.  Thanks again @gabm.  Merging.

**Pull request merged**